--!strict

export type Race =
	"Human"
	| "Namekian"
	| "Acrosian"
	| "Majin"
	| "Saiyan"
	| "Sphinx"
	| "LegendarySuperSaiyan"
	| "ANGEL"
	| "GOD"
	| "KAI"

local Names: { [string]: { string } } = {
	ANGEL = {
		"Cointre",
		"Amare",
		"Kahlua",
		"Sambu",
		"Vermou",
		"Absin",
		"Sherri",
		"Gineba",
		"Vino",
		"Curaca",
		"Cognac",
		"Mezca",
		"Rhum",
		"Bullei",
		"Prose",
		"Mosc",
		"Apero",
		"Grapp",
		"Sake",
		"Char",
	},
	GOD = {
		"Vod",
		"Teq",
		"Bourbo",
		"Scot",
		"Stout",
		"Arak",
		"Pisco",
		"Ouzo",
		"Raki",
		"Cider",
		"Moon",
		"Porter",
		"Ale",
		"Grog",
		"Malbec",
		"Nigori",
		"Bacard",
		"Jäger",
		"Stell",
		"Budo",
	},
	KAI = {
		"Avoca",
		"Litchi",
		"Mago",
		"Kinoa",
		"Pome",
		"Ceras",
		"Saffro",
		"Chia",
		"Durio",
		"Anana",
		"Datt",
		"Pistac",
		"Gojir",
		"Barle",
		"Ruis",
		"Cacao",
		"Maiz",
		"Kape",
		"Lenti",
		"Melo",
	},
	Acrosian = {
		"Glaze",
		"Chill",
		"Frigid",
		"Sub-Zero",
		"Hail",
		"Slush",
		"Cryo",
		"Gelid",
		"Perma",
		"Hoar",
		"Zero",
		"Blizz",
		"Sleet",
		"Rime",
		"Crisp",
		"Algid",
		"Polar",
		"Brumal",
		"Kalt",
		"Frio",
	},
	Human = {
		"Pasta",
		"Noodle",
		"Rice",
		"Stew",
		"Curry",
		"Bento",
		"Miso",
		"Gyoza",
		"Ramen",
		"Udon",
		"Soba",
		"Katsu",
		"Tempu",
		"Nori",
		"Ginger",
		"Basil",
		"Clove",
		"Saffron",
		"Pesto",
		"Olive",
	},
	Majin = {
		"Hocus",
		"Pocus",
		"Kazoo",
		"Presto",
		"Alaka",
		"Zaza",
		"Voo",
		"Juju",
		"Mojo",
		"Bana",
		"Taffy",
		"Mochi",
		"Soufflé",
		"Ganache",
		"Fondue",
		"Gelato",
		"Bakla",
		"Toffee",
		"Nuga",
		"Choco",
	},
	Namekian = {
		"Escal",
		"Lymat",
		"Nerita",
		"Hali",
		"Murex",
		"Oboe",
		"Clar",
		"Bas",
		"Reed",
		"Kora",
		"Sistrum",
		"Lute",
		"Marim",
		"Corno",
		"Helix",
		"Bulim",
		"Cary",
		"Achat",
		"Pom",
		"Vibra",
	},
	Saiyan = {
		"Artich",
		"Toma",
		"Parsley",
		"Celer",
		"Peppar",
		"Zukini",
		"Chard",
		"Kale",
		"Lettu",
		"Okra",
		"Raddis",
		"Broc",
		"Spin",
		"Yamo",
		"Leek",
		"Aspar",
		"Shallot",
		"Tatara",
		"Mani",
		"Kabb",
	},
	LegendarySuperSaiyan = {
		"Brocco",
		"Cassav",
		"Rutab",
		"Daikon",
		"Yamato",
		"Russet",
		"Mallow",
		"Jicamo",
		"Taro",
		"Gourd",
		"Radic",
		"Chokey",
		"Collard",
		"Romanes",
		"Kohl",
		"Parsh",
		"Beet",
		"Turnip",
		"Maize",
		"Kobu",
	},
	Sphinx = {
		"Chivas",
		"Kahlua",
		"Soju",
		"Bourbon",
		"Grappa",
		"Mezcal",
		"Sambuca",
		"Pisco",
		"Brandy",
		"Malibu",
		"Cider",
		"Porter",
		"Stout",
		"Pilsner",
		"Koji",
		"Absin",
		"Tequi",
		"Vod",
		"Gin",
		"Sherry",
	},
}

local function get(race: string): { string }
	return Names[race] or {}
end

return {
	Names = Names,
	get = get,
}
